code/game/machinery/keypads.dm

```diff
@@ -0,0 +1,283 @@
+//Wallframes
+/obj/item/wallframe/keypad
+	name = "keypad frame"
+	desc = "Used for building keypads."
+	icon_state = "keypad_frame"
+	result_path = /obj/machinery/keypad
+	materials = list(MAT_METAL=MINERAL_MATERIAL_AMOUNT)
+	var/password = ""
+
+/obj/item/wallframe/keypad/attack_self(mob/user)
+	if(!user)
+		return 0
+	if(!password)
+		var/pass = input(user,"Please input a new password for this keypad.","Set Password") as null|num
+		if(!in_range(src, user) && loc != user)
+			return
+		if(!user.canUseTopic(src))
+			return
+		if(!pass)
+			user << "<span class='warning>Invaid password!</span>"
+			return
+		pass = round(pass)
+		if(!isnum(pass))
+			user << "<span class='warning>The password must be a number!</span>"
+			return
+		pass = num2text(pass)
+		if(length(pass) > 6)
+			user << "<span class='warning>The password cannot be longer than 6 digits!</span>"
+			return
+		password = pass
+		user << "<span class='notice'>You have set \the [src]'s password to [password].</span>"
+	else
+		user << "<span class='warning'>A password is already set for [src]!</span>" //To fix this, you need a multitool.
+		return
+
+/obj/item/wallframe/keypad/attackby(obj/item/W, mob/user, params)
+	if(istype(W, /obj/item/device/multitool))
+		user.visible_message("<span class='warning'>[user.name] starts resetting the password on \the [src].</span>", \
+							"<span class='notice'>You start resetting the password on \the [src] frame...</span>")
+		playsound(src.loc, 'sound/effects/pop.ogg', 50, 1)
+		if(do_after(user, 20, target = src))
+			user.visible_message("<span class='warning'>[user.name] resets the password on \the [src]!</span>", \
+								"<span class='notice'>You reset the password on \the [src] frame!</span>")
+			playsound(src.loc, 'sound/items/Screwdriver2.ogg', 50, 1)
+			password = ""
+		return
+	..()
+
+/obj/item/wallframe/keypad/attach(turf/on_wall) //I hate how I have to copy-paste code just to prevent the wallframe from being deleted before actually transferring variables.
```

Comment:
Make the main attach return the new object. Then just do var/pass = password, var/obj/machinery/keypad/K = ..(), K.password = pass no?